src/System.Net.Http/src/System/Net/Http/Managed/AuthenticateAndRedirectHandler.cs

```diff
@@ -0,0 +1,223 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Diagnostics;
+using System.Net.Http.Headers;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace System.Net.Http
+{
+    internal sealed partial class AuthenticateAndRedirectHandler : HttpMessageHandler
+    {
+        private readonly HttpMessageHandler _innerHandler;
+        private readonly bool _preAuthenticate;
+        private readonly ICredentials _credentials;
+        private readonly bool _allowRedirect;
+        private readonly int _maxAutomaticRedirections;
+        private AuthenticationHelper.DigestResponse _digestResponse;
```

Comment:
_digestResponse shouldn't be cached at this level, the same handler can send requests to multiple domains. _digestResponse should be cached per domain if at all.

Reply:
The credentials set on the handler is the credential that's going to be used for all requests on this handler. These credentials could be different for different realms on the server, one doesn't know which realm the resource is in by the request. Once the request is sent, the server may request a digest auth for a given realm, and the handler will use that realm information to construct the digestresponse. So it is meaningless to store digestresponse for a realm, as we will not get to know the realm information by looking at the request. This information is obtained from the first not authenticated response from server. Also note, this cached digestresponse is used only when preauthenticate option is set, so I don't see any value in maintaining a mapping between cached digestresponse and realm.

Reply:
I was not referring to realm. You can cache it based on the request prefix. e.g. scheme://host:port

Reply:
Why? Given the digestresponse is used only when preauthenticate is set, i would think it is the responsibility of the user to set it correctly. If the wrong digestresponse is sent, the result is only unauthorized, same as when no digestresponse is sent. Is there any value is extra allocation here?

Reply:
Note the digestresponse here is not the string that goes on the wire, that value is re-computed. The digestresponse here just stores the digest challenge from the server on a prior response.